src/halmos/__main__.py

```diff
@@ -393,38 +403,51 @@ def run(
 
     # check assertion violations
     normal = 0
-    models = []
+    execs_to_model = []
+    models: List[ModelWithContext] = []
     stuck = []
+
     for idx, ex in enumerate(exs):
         if args.debug: print(f'Checking output: {idx+1} / {len(exs)}')
 
         opcode = ex.current_opcode()
         if opcode in [EVM.STOP, EVM.RETURN]:
             normal += 1
         elif opcode in [EVM.REVERT, EVM.INVALID]:
-            # Panic(1) # bytes4(keccak256("Panic(uint256)")) + bytes32(1)
-            if ex.output == int('4e487b71' + '0000000000000000000000000000000000000000000000000000000000000001', 16): # 152078208365357342262005707660225848957176981554335715805457651098985835139029979365377
-                gen_model(args, models, idx, ex)
+            # Panic(1)
+            # bytes4(keccak256("Panic(uint256)")) + bytes32(1)
+            if args.debug: print(f'  Will generate model for Panic(1)')
+            if ex.output == 0x4e487b710000000000000000000000000000000000000000000000000000000000000001:
+                execs_to_model.append((idx, ex))
         elif ex.failed:
-            gen_model(args, models, idx, ex)
+            if args.debug: print(f'  Will generate model for failed execution')
+            execs_to_model.append((idx, ex))
         else:
             stuck.append((opcode, idx, ex))
 
-    end = timer()
+    if len(execs_to_model) > 1 and args.solver_parallel:
+        with Pool(processes=args.solver_parallel_cores) as pool:
+            if args.debug: print(f'Spawning {len(execs_to_model)} parallel assertion solvers on {args.solver_parallel_cores} cores')
+            models = [m for m in pool.starmap(gen_model_from_sexpr, [(args, idx, ex.solver.sexpr()) for idx, ex in execs_to_model])]
 
-    passed = (normal > 0 and len(models) == 0 and len(stuck) == 0)
-    if passed:
-        passfail = color_good('[PASS]')
     else:
-        passfail = color_warn('[FAIL]')
+        models = [gen_model(args, idx, ex) for idx, ex in execs_to_model]
+
+    end = timer()
+
+    no_counterexample = all(m.model is None for m in models)
+    passed = (no_counterexample and normal > 0 and len(stuck) == 0)
+    passfail = color_good('[PASS]') if passed else color_warn('[FAIL]')
```

Comment:
`no_counterexample` should check if all results are unsat, rather than all models are None.

as i mentioned earlier, e.g., if we have a single normal path, and a single panic(1) path for which counterexample generation failed due to timeout, then we shouldn't consider it as passed.

Reply:
I kinda disagree on this one, as a user of halmos I don't want to see a test `[FAIL]` without a valid counterexample I can work with. Maybe a warning that I should increase the assert solver timeout, but it's not a conclusive fail. Which is fine since we're doing bounded model checking, no?

Reply:
make sense. warning would be a better way to convey the result. in that case, i think it would be useful to provide the unsolved queries so that users can directly check them (potentially modifying the queries and/or using different solvers) without re-runing the entire tests.

for this PR, simple warning messages would suffice. we can work on providing queries later.